### PR TITLE
Quarterdeck verify endpoint

### PR DESCRIPTION
**Pull request by pdeziel**
### Scope of changes

This adds an email verification endpoint to Quarterdeck to allow users to verify their email.

Fixes SC-14159

### Type of change

- [x] new feature
- [ ] bug fix
- [ ] documentation
- [ ] testing
- [ ] technical debt
- [ ] other (describe)

### Acceptance criteria

Please verify that the error responses make sense, particularly the case where the token is expired and we have to issue a new one.

### Author checklist

- [x] I have manually tested the change and/or added automation in the form of unit tests or integration tests
- [ ]  I have updated the dependencies list
- [ ]  I have recompiled and included new protocol buffers to reflect changes I made
- [ ]  I have added new test fixtures as needed to support added tests
- [x]  Check this box if a reviewer can merge this pull request after approval (leave it unchecked if you want to do it yourself)
- [x]  I have moved the associated Shortcut story to "Ready for Review"

### Reviewer(s) checklist

- [ ] Any new user-facing content that has been added for this PR has been QA'ed to ensure correct grammar, spelling, and understandability.
- [ ] Are there any TODOs in this PR that should be turned into stories?

### CODE REVIEW

File: pkg/quarterdeck/auth.go [hunk 1]
```diff
@@ -448,3 +449,96 @@ func (s *Server) Refresh(c *gin.Context) {
...
+	if req.Token == "" {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("missing token in request"))
+		return
+	}
+
+	// Look up the user by the token
+	var user *models.User
+	if user, err = models.GetUserByToken(c, req.Token); err != nil {
+		if errors.Is(err, models.ErrNotFound) {
+			c.JSON(http.StatusUnauthorized, api.ErrorResponse("invalid token"))
```
Comment (bbengfort) [inline pkg/quarterdeck/auth.go:479]:
```suggestion
			c.JSON(http.StatusBadRequest, api.ErrorResponse("invalid token"))
```

I think a 400 is probably slightly better here than a 401 for security purposes.

File: pkg/quarterdeck/auth.go [hunk 2]
```diff
@@ -448,3 +449,96 @@ func (s *Server) Refresh(c *gin.Context) {
...
+			}
+
+			// Send the new token to the user
+			s.tasks.Queue(tasks.TaskFunc(func(ctx context.Context) {
+				if err := s.SendVerificationEmail(user); err != nil {
+					log.Error().Err(err).Str("user_id", user.ID.String()).Msg("could not send verification email to user")
+				}
+			}))
+
+			c.JSON(http.StatusGone, api.ErrorResponse("token unavailable"))
```
Comment (bbengfort) [inline pkg/quarterdeck/auth.go:527]:
```suggestion
			c.JSON(http.StatusGone, api.ErrorResponse("token expired, a new verification token has been sent to the email associated with the account"))
```

File: pkg/quarterdeck/auth.go [hunk 3]
```diff
@@ -448,3 +449,96 @@ func (s *Server) Refresh(c *gin.Context) {
...
+			return
+		}
+
+		c.JSON(http.StatusUnauthorized, api.ErrorResponse("invalid token"))
```
Comment (bbengfort) [inline pkg/quarterdeck/auth.go:531]:
```suggestion
		c.JSON(http.StatusBadRequest, api.ErrorResponse("invalid token"))
```

File: pkg/quarterdeck/db/errors.go [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package db
+
+import "errors"
+
+var (
+	ErrTokenMissingEmail = errors.New("token is missing email address")
+	ErrTokenExpired      = errors.New("token has expired")
+	ErrInvalidSecret     = errors.New("invalid secret for token verification")
+	ErrTokenInvalid      = errors.New("token has invalid signature")
```
Comment (bbengfort) [inline pkg/quarterdeck/db/errors.go:9]:
```suggestion
	ErrTokenMissingEmail = errors.New("email verification token is missing email address")
	ErrTokenExpired      = errors.New("email verification token has expired")
	ErrInvalidSecret     = errors.New("invalid secret for email token verification")
	ErrTokenInvalid      = errors.New("email verification token has invalid signature")
```

We have a lot of tokens in Quarterdeck; may help debugging if we specify.

File: pkg/quarterdeck/auth_test.go [hunk 5]
```diff
@@ -302,3 +302,53 @@ func (s *quarterdeckTestSuite) TestRefresh() {
...
+	// Test that an empty token is rejected
+	err := s.client.VerifyEmail(ctx, &api.VerifyRequest{})
+	s.CheckError(err, http.StatusBadRequest, "missing token in request")
+
+	// Test that an error is returned if it doesn't exist in the database
+	req := &api.VerifyRequest{
+		Token: "wrongtoken",
+	}
+	err = s.client.VerifyEmail(ctx, req)
+	s.CheckError(err, http.StatusUnauthorized, "invalid token")
```
Comment (bbengfort) [inline pkg/quarterdeck/auth_test.go:320]:
```suggestion
	s.CheckError(err, http.StatusBadRequest, "invalid token")
```

File: pkg/quarterdeck/db/models/users.go [hunk 6]
```diff
@@ -51,6 +51,7 @@ type User struct {
 const (
 	getUserIDSQL    = "SELECT name, email, password, terms_agreement, privacy_agreement, email_verified, email_verification_expires, email_verification_token, email_verification_secret, last_login, created, modified FROM users WHERE id=:id"
 	getUserEmailSQL = "SELECT id, name, password, terms_agreement, privacy_agreement, email_verified, email_verification_expires, email_verification_token, email_verification_secret, last_login, created, modified FROM users WHERE email=:email"
+	getUserTokenSQL = "SELECT id, name, email, password, terms_agreement, privacy_agreement, email_verified, email_verification_expires, email_verification_secret, last_login, created, modified FROM users WHERE email_verification_token=:token"
```
Comment (bbengfort) [inline pkg/quarterdeck/db/models/users.go:54]: Should we omit the password when selecting the user details for a token? It's not needed for this endpoint and reducing the amount of time even a hashed password in memory is probably a good idea.

File: pkg/quarterdeck/db/tokens.go [hunk 7]
```diff
@@ -111,7 +111,7 @@ func (t *VerificationToken) Verify(signature string, secret []byte) (err error)
 
 	// Check if the recomputed token matches the user token
 	if !hmac.Equal(mac.Sum(nil), token) {
-		return errors.New("invalid token signature")
+		return ErrTokenInvalid
```
Comment (bbengfort) [inline pkg/quarterdeck/db/tokens.go:114]: Nice!

**Review by bbengfort — APPROVED** [review]
Looks great! Getting close to this feature being ready!

File: pkg/quarterdeck/db/models/users.go (hunk 6 above)
Comment (pdeziel) [reply pkg/quarterdeck/db/models/users.go:54]: Good idea

Comment (pdeziel) [reply pkg/quarterdeck/db/models/users.go:54]: But the problem is that the db `Save()` method on the user requires the entire user object.

Comment (bbengfort) [reply pkg/quarterdeck/db/models/users.go:54]: Hmm … ok; I'll create a technical debt story for this.

**Pull request merged**